scripts/adapters/update-copyright-headers.rb

```diff
@@ -11,17 +11,52 @@
 # Iterate over all the files in the source directory
 for_all_source_files do |file_path, contents|
   # Skip if file already has the copyright header
-  if contents.start_with?(SOURCE_FILE_COPYRIGHT_NOTICE)
+  if contents.match(SOURCE_FILE_COPYRIGHT_NOTICE_CURRENT_YEAR_ONLY_REGEX)
+    # No op if the copyright year is the CURRENT year
     next
+  elsif contents.match(SOURCE_FILE_COPYRIGHT_NOTICE_FROM_AND_CURRENT_YEARS_REGEX)
+    # No op if the copyright year is from 20XX to the CURRENT year
+    next
+  elsif contents.match(SOURCE_FILE_COPYRIGHT_NOTICE_FROM_YEAR_ONLY_REGEX)
+    # The file starts with copyright header that contains outdated FROM year
+    tmp = Tempfile.new("tmp")
+    begin
+      currentYearStartIndex = SOURCE_FILE_COPYRIGHT_NOTICE_TIME_SPAN_DASH_INDEX + 1
+      contents.insert(SOURCE_FILE_COPYRIGHT_NOTICE_TIME_SPAN_DASH_INDEX, "-#{Time.now.year}") # add the current year
+      tmp.puts contents
+
+      # Replace the original file with the new one
+      FileUtils.mv(tmp.path, file_path)
+      modified_files.append(file_path)
+    ensure
+      tmp.close
+      tmp.unlink
+    end
+  elsif contents.match(SOURCE_FILE_COPYRIGHT_NOTICE_FROM_AND_TO_YEARS_REGEX)
+    # The file starts with copyright header that contains outdated TO year
+    tmp = Tempfile.new("tmp")
+    begin
+      currentYearStartIndex = SOURCE_FILE_COPYRIGHT_NOTICE_TIME_SPAN_DASH_INDEX + 1
+      contents[currentYearStartIndex..currentYearStartIndex+3] = "#{Time.now.year}" # replace the outdated TO year with the current year
+      tmp.puts contents
+
+      # Replace the original file with the new one
+      FileUtils.mv(tmp.path, file_path)
+      modified_files.append(file_path)
+    ensure
+      tmp.close
+      tmp.unlink
+    end
   else
+    # The file does not start with the copyright header
     tmp = Tempfile.new("tmp")
     begin
       if contents.start_with?("//")
-          # If the file starts with a comment block, replace it with the copyright header
-          tmp.puts contents.sub(COMMENT_BLOCK_AT_BEGINNING_OF_FILE_REGEX, SOURCE_FILE_COPYRIGHT_NOTICE)
+        # If the file starts with a comment block, replace it with the copyright header assuming this file is created this year
```

Comment:
This brings up a good question: should all files in the repo have the same from year? If so, this won't work, as this would just add the current year. For example, let's say our Mediation adapters should be `2022-2023`. When we create a new file in a Mediation adapter, this would just add `2023` only, and when 2024 rolls around, it would become `2023-2024`, when it should be `2022-2024`.

To fix this, we'd actually need to be told externally what the from year is, e.g. by passing the from year in from the calling workflow (via param or env variable), or having the from year determined from elsewhere in the repo, or something like that.

Reply:
Maybe we can use the CHARTBOOST_PLATFORM environment variable to determine what the from year is.

Reply:
We could do that, but if we then create a new Mediation adapter in 2024, should that share the same start year as every other Mediation adapter?

This might be worth bringing up with product to see what they think.

Reply:
@pleasesavemycat said we should use the same date (or date range) per repo, but each repo should use its own date (or date range), so we'll probably need to specify either a workflow variable, or an environment variable that's set in the repo.

Reply:
Maybe we could fetch the year by reading the date of the first comment in the repo.

Reply:
This seems like it would create a chicken and egg problem, where the script would fix header comments, but would require at least one to be in the correct format to know how to fix all of the others.

An action or env var is a bit more painful now, but it fixes the issue without any unexpected behavior like this.

Reply:
Sorry I mistyped. I meant reading the first **commit** in the repo.

Reply:
lol that was me

Reply:
Hmm that's an interesting idea, if feasible. Maybe we could use the GitHub API for that?

Reply:
Yeah it should be possible with the GitHub API

Reply:
Learned from a few Stackoverflow articles, we can get the YEAR of the first commit with:
1. `git rev-list --max-parents=0 HEAD` // the hash of the first commit in the repo
2. `git show -s --format=%cd --date=format:%Y <commit>` // show the YEAR of a given commit
3. `command1 | xargs -I{} command2 {}` // pass the result of the 1st command to the 2nd command as the argument
4. With all above combined: `git rev-list --max-parents=0 HEAD | xargs -I{} git show -s --format=%cd --date=format:%Y {}`

As mentioned earlier, we won't need to define any repo specific env var for the starting year. 

There is one concern, though. The first commit of the Mediation repo was made in 2018, but we current headers starts from 2022. Should we respect 2018?

Reply:
These scripts do not affect the helium repo though, only adapters, so we should be fine using the date from the first commit?
At some point when we open-source everything we could have the adapter actions repo depend on the common scripts repo. Then we could use an env variable and fallback to the first commit date when it's not defined.

Reply:
> These scripts do not affect the helium repo though, only adapters

I wasn't sure this is the best approach so I didn't make PRs for all places. 

Just posted a Slack question about why it Helium iOS copyright header started in 2022 but not 2018
https://chartboost.slack.com/archives/C019BTYC9MH/p1696281292875779